### DISCUSSION TITLE
What frameworks are available and what are the pros and cons

### DISCUSSION THREAD

**Discussion by MikeNeilson** [opendcs/rest_api#4] [Ideas]
# REST API frameworks

| framework | link | pros | cons|
|-------------|-----|------|-----|
|JAX-RS |https://github.com/jakartaee/rest| <ul><li>Jakarta (nee Java) EE standard</li><li> multiple implementations</li><li>container (Tomcat/Jetty/Jboss/etc) agnostic</li></ul>| Likely slow to implement new changes due to standards process|
|Javalin|https://javalin.io/| <ul><li>Simple</li><li>Hooks for various Auth mechanism</li><li>good Swagger (OpenAPI) doc generation</li><li>Designed to run standalone or within an App server</li><li>Used by an OpenDCS partner agency (https://github.com/USACE/cwms-radar-api)</li></ul>| There is some initial setup for validation.|
|Spring|spring.io|Has a library for everything|Has a library for everything|

# SQL Frameworks
|framework|link|pros|cons|license|active developers|
|-----------|-----|-----|-----|-------|------|
|JDBC|<built in>|It'll always be there.|Provides little help for type checking. Lots of manual work.|depends on provide|yes|
|JDBI|https://jdbi.org/|<ul><li>Completely OpenSource</li><li>Lite weight wrapper around JDBC</li><li>Provides type checking</li><li>Uses java 8+ streams and groupby which is good for joins</li><li>Abstracts SQL syntax issues for the caller (for example returning the generated ID without making another query)</li></ul>|Still some manual work for type checking to behave correctly, but less than JDBC.|Apache-2.0|yes|
|SqlQuery|http://querydsl.com/|<ul><li>Code generation from active db to enforce type checking</li><li>Abstracts Database differences</li></ul>|<ul><li>Code generation -- we support 3 databases, one already uses JOOQ</li></ul>|Apache-2.0|Yes, however it did have some previous downtime|
|JOOQ|jooq.org|<ul><li>Code generation</li><li>Abstracts Database differences</li></ul>|<ul><li>Code generation</li><li>Only free for the open source databases (SQLITE,Postgres, MariaDB, etc)</li></ul>|OpenSource/Commerical|very|
|JPA (hibernate,spring,etc)||well supported|Basically no control over the SQL without a lot of work. This would likely be problematic in OpenDCS as we support 3 independant databases. At least within the current design. Would require altering every current class that holds database data with annotations or creating new classes to transfer data.|various|yes|

**Comment by MikeNeilson** [comment]
@opendcs/opendcs-core-devs

**Comment by agilmore2** [comment]
The only criteria I'd add to the usual is a framework that Reclamation
Enterprise IT folks are willing to host. Not sure we could get an answer to
it, but that's what I'm concerned about.

**Reply by MikeNeilson** [reply]
The target is to support running in one of the many Java EE Applications servers:

- Tomcat
- TomEE
- Jboss
- Glassfish
- Jetty (standalone but can proxy through apache or haproxy if they have concerns.)

I'd be surprised though if literally none of those were supported, if you're using Apex from ORDS you've got Tomcat going.

Granted an initial release might have some problems meeting that goal being a rather broad scope but the final target will be "the Web profile" and not a specific Application server. (at least as far as continued development by the OpenDCS group goes.)

**Comment by adamkorynta** [comment]
I can vouch for jOOQ having excellent documentation.

**Reply by MikeNeilson** [reply]
very, though the licensing makes it difficult, it would have to be an only for the reference postgresql database and the rest is onus on the agency implementation. Maybe that's a good thing, I'm not sure. I feel like JDBI would be a reasonable balance though with letting some default implementation exist for things like the Enums, etc that don't really change between databases.

**Reply by adamkorynta** [reply]
Yeah and JDBI's documentation looks good, though I can't vouch for it in practice.

**Comment by krowvin** [comment]
Might be a good place to list frontend frameworks as well? Current and planned.

**Reply by MikeNeilson** [reply]
Wasn't worried about it at the time, but definitely a good thing to get in now. If you have some thoughts go ahead and post them.
